tests/active_poller.cpp

```diff
@@ -0,0 +1,414 @@
+{
+    /// \todo is it good that this is accepted? should probably already be 
+    ///   checked by zmq_poller_add/modify in libzmq:
+    ///   https://github.com/zeromq/libzmq/issues/3088
+    zmq::context_t context;
+    zmq::socket_t socket{context, zmq::socket_type::router};
+    zmq::active_poller_t active_poller;
+    zmq::active_poller_t::handler_t handler;
+    short invalid_events_type = 2 << 10;
+    ASSERT_NO_THROW(active_poller.add(socket, invalid_events_type, handler));
+    ASSERT_FALSE(active_poller.empty ());
+    ASSERT_EQ(1u, active_poller.size ());
+}
+
+TEST(active_poller, add_handler_twice_throws)
+{
+    zmq::context_t context;
+    zmq::socket_t socket{context, zmq::socket_type::router};
+    zmq::active_poller_t active_poller;
+    zmq::active_poller_t::handler_t handler;
+    active_poller.add(socket, ZMQ_POLLIN, handler);
+    /// \todo the actual error code should be checked
+    ASSERT_THROW(active_poller.add(socket, ZMQ_POLLIN, handler), zmq::error_t);
+}
+
+TEST(active_poller, wait_with_no_handlers_throws)
+{
+    zmq::active_poller_t active_poller;
+    /// \todo the actual error code should be checked
+    ASSERT_THROW(active_poller.wait(std::chrono::milliseconds{10}), zmq::error_t);
+}
+
+TEST(active_poller, remove_unregistered_throws)
+{
+    zmq::context_t context;
+    zmq::socket_t socket{context, zmq::socket_type::router};
+    zmq::active_poller_t active_poller;
+    /// \todo the actual error code should be checked
+    ASSERT_THROW(active_poller.remove(socket), zmq::error_t);
+}
+
+TEST(active_poller, remove_registered_empty)
+{
+    zmq::context_t context;
+    zmq::socket_t socket{context, zmq::socket_type::router};
+    zmq::active_poller_t active_poller;
+    active_poller.add(socket, ZMQ_POLLIN, zmq::active_poller_t::handler_t{});
+    ASSERT_NO_THROW(active_poller.remove(socket));
+}
+
+TEST(active_poller, remove_registered_non_empty)
+{
+    zmq::context_t context;
+    zmq::socket_t socket{context, zmq::socket_type::router};
+    zmq::active_poller_t active_poller;
+    active_poller.add(socket, ZMQ_POLLIN, [](short) {});
+    ASSERT_NO_THROW(active_poller.remove(socket));
+}
+
+namespace {
+struct server_client_setup : common_server_client_setup
+{
+    zmq::active_poller_t::handler_t handler = [&](short e) {
+        events = e;
+    };
+
+    short events = 0;
+};
+}
+
+TEST(active_poller, poll_basic)
+{
+    server_client_setup s;
+
+    ASSERT_NO_THROW(s.client.send("Hi"));
+
+    zmq::active_poller_t active_poller;
+    bool message_received = false;
+    zmq::active_poller_t::handler_t handler = [&message_received](short events) {
+        ASSERT_TRUE(0 != (events & ZMQ_POLLIN));
+        message_received = true;
+    };
+    ASSERT_NO_THROW(active_poller.add(s.server, ZMQ_POLLIN, handler));
+    ASSERT_EQ(1, active_poller.wait(std::chrono::milliseconds{-1}));
+    ASSERT_TRUE(message_received);
+}
+
+/// \todo this contains multiple test cases that should be split up
```

Comment:
can we remove this comment?

Reply:
Do you think it no longer applies?

Reply:
I think it has never applied. All separate cases are already covered? This is a test that exercises more complex scenario - splinting it up does not make sense because you loose the indent of it.

Reply:
Ok, then remove it.